Add missing mania tooltip overlay for 4k and 7k

NicholasChin28 opened the pull request:
Attempts to fix #30827 by adding osu!mania 4K/7K rank breakdowns to profile overlay tooltips, matching the behavior on the osu! website.

Changes:
- Added FourKey and SevenKey to CommonStrings for localization support
- Implemented Variant mapping in UserStatistics to properly handle 4K/7K data from the API 
- Added variant rank display logic in MainDetails for both global and country rank tooltips

The implementation follows the web version's data structure, using the variants field from the API response to show individual rankings for each key mode.

Testing:
- Hover over global rank in osu!mania profile to see 4K/7K breakdown
- Hover over country rank to see corresponding 4K/7K country rankings

NicholasChin28 committed:
Add missing mania tooltip overlay for 4k and 7k

Joehuu commented on osu.Game/Localisation/CommonStrings.cs:
```diff
@@ -179,6 +179,16 @@ public static class CommonStrings
+        public static LocalisableString FourKey => new TranslatableString(getKey(@"four_key"), @"4K");
+
+        /// <summary>
+        /// "7K"
```
There's already strings in web:
https://github.com/ppy/osu-resources/blob/5642a9c5f18b5c17a2f7508f12f295524b976a71/osu.Game.Resources/Localisation/Web/BeatmapsStrings.cs#L1207-L1215

Joehuu commented on osu.Game/Users/UserStatistics.cs:
```diff
@@ -118,5 +125,38 @@ public int this[ScoreRank rank]
+            [JsonConverter(typeof(StringEnumConverter))]
+            public GameVariant? VariantType;
+
+            public LocalisableString VariantDisplay => VariantType switch
+            {
+                GameVariant.FourKey => CommonStrings.FourKey,
+                GameVariant.SevenKey => CommonStrings.SevenKey,
+                _ => string.Empty
+            };
```
This can just be a `LocalisableDescription` attribute in the enum members, with `GetLocalisableDescription()`.

NicholasChin28 committed:
Replace switch expression with LocalisableDescription attribute for variant display

Use existing localisation strings from BeatmapsStrings instead of CommonStrings for consistent localisation handling

Joehuu reviewed (commented):
Forgot to mention this, but contributors are expected to fix the code quality issues listed below in the checks.

The anchor doesn't seem to match web's tooltip. Can probably apply this to all tooltips:
```diff
diff --git a/osu.Game/Graphics/Cursor/OsuTooltipContainer.cs b/osu.Game/Graphics/Cursor/OsuTooltipContainer.cs
index 0d36cc1d08..4180825a8d 100644
--- a/osu.Game/Graphics/Cursor/OsuTooltipContainer.cs
+++ b/osu.Game/Graphics/Cursor/OsuTooltipContainer.cs
@@ -80,6 +80,7 @@ public OsuTooltip()
                         Margin = new MarginPadding(5),
                         AutoSizeAxes = Axes.Both,
                         MaximumSize = new Vector2(max_width, float.PositiveInfinity),
+                        TextAnchor = Anchor.TopCentre,
                     }
                 };
             }

```

as web also defaults to `TopCentre` everywhere.
<img width="310" alt="Screenshot 2024-12-12 at 1 03 15 PM" src="https://github.com/user-attachments/assets/930ae97f-9f7e-4ccd-bf09-3b7a7ffc18aa" />

NicholasChin28 committed:
Fix text anchor for mania tooltip

NicholasChin28 commented:
I have added the text anchor for the tooltip

NicholasChin28 committed:
Use Count comparison instead of Any

NicholasChin28 committed:
Fix code quality issues

bdach committed:
Clean up model

- Properly annotate things as nullable
- Remove weird passthrough property (more on that later)

bdach committed:
Split actual methods & fix completely broken localisation

Localisable strings cannot be plainly interpolated or joined. That is a
lossy operation that loses data.

bdach committed:
Revert "Fix text anchor for mania tooltip"

This reverts commit c0b6e784a5076dbaf6addbfdae00bdebd35c3f6f.

The change affects editor and other stuff and I'm not sure it's correct.
It's not like client needs to match the appearance really. It already
doesn't in many places.

bdach commented:
I've done things to get this in. Notably the model was wonky (some things nullable that shouldn't have been, and some things non-nullable that should have been) and the localisation was broken.

I've also reverted the tooltip change as I find it disputable and I don't think it matters enough anyway because the client design is already not matching web in several respects.
